### PR TITLE
feat/AB#62230-implement-autocomplete

### PR DESCRIPTION
**Pull request by estelafs**
# Description
Created autocomplete directive to the new UI library with:
- Options list can be grouped
- Options displayed are filtered when typing in the input

## Ticket
[ABC- Implement Autocomplete component](https://dev.azure.com/WHOHQ/EMSSAFE/_workitems/edit/62230)

## Type of change
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
Stories files were served and checked

## Sreenshots
Autocomplete panel inject in a input with simples (not grouped) options:
simple-autocom

Autocomplete panel inject in a input with grouped option:
grouped-autocom

Autocomplete with form control:
form-autocom

# Checklist:

( * == Mandatory ) 

- [X] * My code follows the style guidelines of this project
- [X] * Linting does not generate new warnings
- [X] * I have performed a self-review of my own code
- [X] * I have commented my code, particularly in hard-to-understand areas
- [X] * I have put JSDoc comment in all required places
- [X] * My changes generate no new warnings
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have included screenshots describing my changes if relevant
- [X] * I have selected labels in the Pull Request, according to the changes with code brings
- [ ] I have made corresponding changes to the documentation ( if required )
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: libs/ui/src/lib/autocomplete/autocomplete.component.stories.ts [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+  const formGroup = new FormGroup({
+    autocomplete: new FormControl(''),
+  });
+  args.options = options;
+  args.required = true;
+  args.placeholder = 'Select a value';
+  return {
+    component: AutocompleteComponent,
+    template: `
+      <form [formGroup]="formGroup">
```
Comment (unai-reliefapp) [inline libs/ui/src/lib/autocomplete/autocomplete.component.stories.ts]: Please double check how the material automplete is implemented:
image
It's using a directive to link the autocomplete list to the needed input. Please follow the same guide, so we can inject the autocomplete list to any input, regarding of it's styling and methods.

You have an example of how we link a directive to another element in here: https://github.com/ReliefApplications/oort-frontend/blob/feat/AB%2361955_ABC-implement-flyout-menu-component/libs/ui/src/lib/menu/menu.component.stories.ts#L37-L58
The directive used: https://github.com/ReliefApplications/oort-frontend/blob/feat/AB%2361955_ABC-implement-flyout-menu-component/libs/ui/src/lib/menu/menu.directive.ts

File: libs/ui/src/lib/autocomplete/autocomplete.component.html [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  </div>
+</form>
+
+<ng-template #recursiveOptionsTemplate let-list="list">
+  <ul
+    id="autocomplete"
+    *ngFor="let item of list; let i = index"
+    class="z-20 mt-1 p-1 py-1 focus:outline-none sm:text-sm"
+    role="autocomplete" aria-labelledby="autocomplete-label" aria-activedescendant="autocomplete-option-3"
+    [ngClass]="{
```
Comment (unai-reliefapp) [inline libs/ui/src/lib/autocomplete/autocomplete.component.html]: For the hover effect on option background please use the `bg-primary`


Thanks

File: libs/ui/src/lib/autocomplete/autocomplete.component.ts
Comment (unai-reliefapp) [inline libs/ui/src/lib/autocomplete/autocomplete.component.ts:1]: Please double check the selection check logic:
image
image

**Review by unai-reliefapp — CHANGES_REQUESTED** [review]

**Comment by estelafs** [comment]
@unai-reliefapp I updated the PR with the requested changes (in the end the component was not needed, only the directive). Please take a look and let me know if anything is wrong

**Comment by unai-reliefapp** [comment]
> @unai-reliefapp I updated the PR with the requested changes (in the end the component was not needed, only the directive). Please take a look and let me know if anything is wrong

Hi @estelafs , I updated the PR with some changes regarding DOM manipulation using Renderer2 in Angular, that's the correct way of doing it: https://angular.io/api/core/Renderer2

Besides that, two notes:

- with the current implementation what happens if want to access the options children and the accessor is different than `children`, as here for example?

image
Are we going to map all the options given to the directive so they follow the AutocompleteItem interface?

- If there is an icon involved in each item of the autocomplete list as here?

image

The use of a component to handle the option projection was in order to take care of these two points mainly,

Also please look that we are using a custom function input to tell the autocomplete how we want to display the items in the list, as here: https://github.com/ReliefApplications/oort-frontend/blob/tailwind/libs/safe/src/lib/components/ui/tagbox/tagbox.component.html#L40
Please make sure that we can add a custom display to the list items as well if needed

Thanks, any doubt please write me back

**Pull request merged**